debezium/dbz#1417 Add support for quantiles

vjuranek opened the pull request:
Add support for metrics statistics like averages and quantiles.

debezium/dbz#1417

vjuranek committed:
debezium/dbz#1417 Add option to collect statistics

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Add quantiles support using DDSketch

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Use quantiles with lag behind source

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Add lag behind source MX bean

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Add statistics factory

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Add java docs

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Use generic type

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Adjust tests

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Skip statistics tests on SQL server

SQL server metrics don't include lagBehindSource metric, which is used
for testing statistics.

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

Naros commented:
:x: Developer Certificate of Origin (DCO) check failed.

Hi @vjuranek, please sign off all commits with:

```
git commit -s
```

If pull request commits are not signed off, the pull request cannot be merged. For more information about why this is required, please see [our blog](https://debezium.io/blog/2025/12/12/contribution-requirements-changing/) about contribution requirement changes.

mfvitale commented:
@vjuranek just a question. How this work fits with the introduction of `micrometer` that we have planned for Q2?

vjuranek commented:
> @vjuranek just a question. How this work fits with the introduction of `micrometer` that we have planned for Q2?

@mfvitale well, TBH I don't know much about Micrometer and what are the exact plan how to use it in Debezium. This work just extends Debezium metrics. My understanding (based on the Debezium roadmap `Micrometer monitoring using OpenTelemetry`) is that Mictometer will replace OTEL tracing, providing general observability capabilities and OTEL will be only one of the possible implementations. So if this understanding is correct, there is no overlap between these two.

vjuranek committed:
debezium/dbz#1417 Make Measurements thread safe

Also add missing `reset()` method and few other minor fixes to address
review comments.

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Fix type of the average value

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Start and stop measurement collector properly

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek committed:
debezium/dbz#1417 Fix typos and other minor fixes

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek commented on debezium-embedded/src/test/java/io/debezium/pipeline/AbstractMetricsTest.java:
```diff
@@ -275,6 +311,56 @@ protected void assertStreamingMetrics(boolean checkAdvancedMetrics, long expecte
+        // Insert new records and wait for them to become available
+        executeInsertStatements();
+
+        consumeRecordsByTopic((int) expectedEvents);
+        Thread.sleep(Duration.ofSeconds(2).toMillis());
```
This seems to be fine -  we just need to wait little bit to be sure that metrics is update. Moreover, similar approach is used across the whole test.

vjuranek committed:
debezium/dbz#1417 Reset measurement queue on meter reset

Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

mfvitale commented:
> > @vjuranek just a question. How this work fits with the introduction of `micrometer` that we have planned for Q2?
> 
> @mfvitale well, TBH I don't know much about Micrometer and what are the exact plan how to use it in Debezium. This work just extends Debezium metrics. My understanding (based on the Debezium roadmap `Micrometer monitoring using OpenTelemetry`) is that Mictometer will replace OTEL tracing, providing general observability capabilities and OTEL will be only one of the possible implementations. So if this understanding is correct, there is no overlap between these two.

I'm not a super expert of Micrometer but I'm wondering if we could also use it for the metrics too. Micrometer should offers already the concept of quantiles.
So in that case we can avoid to use another dependency to do the calculations. WDYT?

jpechane commented:
@rk3rn3r Could you please take a look at this as IIRC you'll be involved in Micrometer in future?

vjuranek committed:
debezium/dbz#1417 Add sum of the values
Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

vjuranek commented on debezium-connector-common/src/main/java/io/debezium/metrics/stats/SingleValueMeasurement.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author vjuranek
+ */
+public interface SingleValueMeasurement<T extends SingleValueEvent<V>, V> extends Measurement<T> {
```
Sure, added.

vjuranek commented:
> I'm not a super expert of Micrometer but I'm wondering if we could also use it for the metrics too. Micrometer should offers already the concept of quantiles.
> So in that case we can avoid to use another dependency to do the calculations. WDYT?

Yes, I was looking on Mircometer yesterday little bit (and will continue today) and it indeed provides support for histograms and quantiles. However, my understanding is that it's just a facade, so you won't avoid additional dependency on e.g. Datadog lib anyway, but it gives you the freedom to switch the underlying data processing library.

vjuranek commented:
I did some investigation over the weekend and also consult with the Claude. It really depends what out intentions with Micrometer are. If it's limited only to tracing, there is of course no overlap. If we are going to replace all current metrics (JMX) with Micrometer, it's questionable if this PR should be merged. Claude concluded (and I more or less agree with it), that if switch to Micrometer will be done any time soon (in Q2 as planned), this PR shouldn't be merged. If it takes loner time (as it will be bigger effort replace all metrics with Micrometer), it makes sense to merge this PR, eventually add into the PR another abstraction layer, which would make the switch to Micrometer straightforward.

jpechane commented:
@rk3rn3r Could you please comment above on the timing?
@vjuranek I don't think we want to remove JMX metrics altogether, as it is universal tool in Java world, so maybe it should still stay?

mfvitale commented:
> I don't think we want to remove JMX metrics altogether, as it is universal tool in Java world, so maybe it should still stay?

@jpechane, @vjuranek Well I think that we should use `Micrometer` as primary instrumentation library and then export the metrics via micrometer `registry` (look to [Composite Registries](https://docs.micrometer.io/micrometer/reference/concepts/registry.html#_composite_registries)) to the different technology (jmx, prometheus, otel)

Think to it as `SLF4J` and its implementations (Logback/Log4j2). 

In that way we use a more powerful instrumentation library that has concepts like `Gauges`, `Counters`, `Timers` and so on and the let the registry adapt them to the destination.

vjuranek commented:
```quote
Well I think that we should use Micrometer as primary instrumentation library and then export the metrics via micrometer registry
```
@mfvitale,  in general I agree and if I was starting a new project, I'd use Micrometer as the primary monitoring/tracing library in such case. On the other hand, if you have quite large project with lots of existing JMX metrics (which doesn't require any dependency)  and switching to Micrometer would require significant effort, I'd be quite cautious. AFAICT Micrometer is not under any foundation, so if they decide to change the license and forbid it for commercial use or do any other disrupting change, we would have to invest once again big effort to switch to another solution or move JMX metics back.

mfvitale commented:
> ```
> Well I think that we should use Micrometer as primary instrumentation library and then export the metrics via micrometer registry
> ```
> 
>  AFAICT Micrometer is not under any foundation, so if they decide to change the license and forbid it for commercial use or do any other disrupting change, we would have to invest once again big effort to switch to another solution or move JMX metics back.

Well this is a good point.

rk3rn3r commented:
Sorry, I talked to Jiri about it last week and wrote my reply and just realized now, that I didn't submit. 

I agreed with Jiri. Ours and my personal take was we should merge the PR. It's possible to start from here and it will move things forward.
If we replace / move to metrics with Micrometer eventually we are still able to decide in H2-26 when we get a feeling for Micrometer.
My personal feeling is that it's easier to decide when you started using it.

jpechane commented on debezium-bom/pom.xml:
```diff
@@ -605,6 +605,13 @@
             </dependency>
 
+            <!-- Statistics dependencies -->
+            <dependency>
```
@vjuranek Are there any transitive dependencies pulled with it?

vjuranek replied:
No transitive dependency (optional dependency on protobuf, which we already include)

jpechane commented:
@vjuranek Can you please add dew screenshots from a JMX console?

vjuranek commented:
@jpechane sorry for the delay, here are the screenshots:

<img width="1220" height="675" alt="Screenshot_20260504_135403" src="https://github.com/user-attachments/assets/9153728c-a205-4747-b896-12cbb21f4526" />
<img width="1220" height="675" alt="Screenshot_20260504_135416" src="https://github.com/user-attachments/assets/502dcfc9-4a1f-4159-9824-5c1ffdda9535" />
<img width="1220" height="675" alt="Screenshot_20260504_135427" src="https://github.com/user-attachments/assets/6d5b11f2-ba7c-42be-97e2-f75b256aa1da" />
<img width="1220" height="675" alt="Screenshot_20260504_135439" src="https://github.com/user-attachments/assets/6637172e-32c2-4456-8a51-0c7758e7b28b" />
<img width="1220" height="675" alt="Screenshot_20260504_135448" src="https://github.com/user-attachments/assets/b0b53c86-9682-4d11-b713-5aa0e14c4d5f" />
<img width="1220" height="675" alt="Screenshot_20260504_135456" src="https://github.com/user-attachments/assets/e375e1c6-cf63-4666-9d4d-0291949e3304" />

vjuranek committed:
debezium/dbz#1417 Don't hardcode number of events in the metric tests
Signed-off-by: Vojtech Juranek <vjuranek@redhat.com>

jpechane reviewed (approved):
@vjuranek LGTM, thanks! Feel free to merge and continue with the changes we discussed during yesterday call.
